plugins/XmlMaterialProfile/XmlMaterialProfile.py

```diff
@@ -3,32 +3,196 @@
 
 import math
 import copy
+import io
 import xml.etree.ElementTree as ET
+import uuid
 
 from UM.Logger import Logger
 
-import UM.Settings
+import UM.Dictionary
 
-# The namespace is prepended to the tag name but between {}.
-# We are only interested in the actual tag name, so discard everything
-# before the last }
-def _tag_without_namespace(element):
-        return element.tag[element.tag.rfind("}") + 1:]
+import UM.Settings
 
+##  Handles serializing and deserializing material containers from an XML file
 class XmlMaterialProfile(UM.Settings.InstanceContainer):
     def __init__(self, container_id, *args, **kwargs):
         super().__init__(container_id, *args, **kwargs)
 
+    ##  Overridden from InstanceContainer
+    def duplicate(self, new_id, new_name = None):
+        result = super().duplicate(self.getMetaDataEntry("brand") + "_" + new_id, new_name)
+        result.setMetaDataEntry("GUID", str(uuid.uuid4()))
+        return result
+
+    ##  Overridden from InstanceContainer
+    def setMetaDataEntry(self, key, value):
+        super().setMetaDataEntry(key, value)
+
+        for container in UM.Settings.ContainerRegistry.getInstance().findInstanceContainers(GUID = self.getMetaDataEntry("GUID")):
+            container.setMetaData(copy.deepcopy(self._metadata))
+
+    ##  Overridden from InstanceContainer
+    def setProperty(self, key, property_name, property_value, container = None):
+        super().setProperty(key, property_name, property_value)
+
+        for container in UM.Settings.ContainerRegistry.getInstance().findInstanceContainers(GUID = self.getMetaDataEntry("GUID")):
+            container._dirty = True
+
+    ##  Overridden from InstanceContainer
     def serialize(self):
-        raise NotImplementedError("Writing material profiles has not yet been implemented")
+        registry = UM.Settings.ContainerRegistry.getInstance()
+
+        all_containers = registry.findInstanceContainers(GUID = self.getMetaDataEntry("GUID"))
+        most_generic = all_containers[0]
+        for container in all_containers:
+            # Find the "most generic" version of this material
+            # This is a bit of a nasty implementation because of the naive assumption that anything with a shorter
+            # id is more generic. It holds for the current implementation though.
+            if len(most_generic.id) > len(container.id):
+                most_generic = container
```

Comment:
After thinking a bit more about it, even with our current implementation doesn't quite hold. If a material specifies a profile for `ultimaker2_plus_0.4` and one for `ultimaker2_extended_plus`, then the profile for `ultimaker2_plus_0.4` will get chosen over `ultimaker2_extended_plus`.

I'd still rather have you checking with multiple filter steps, like we discussed. It will be more code, but it could prevent some hard-to-debug bugs in the future.